File: riveredge-frontend/src/apps/kuaicaiwu/services/finance/document-reconciliation.ts
import { apiRequest } from '../../../../services/api';

export interface DocumentReconciliationGapParams {
  partner_type: 'Customer' | 'Supplier';
  partner_id: number;
  start_date: string;
  end_date: string;
  only_gaps?: boolean;
  keyword?: string;
  doc_type?: string;
  doc_code?: string;
  sort_field?: string;
  sort_order?: string;
  skip?: number;
  limit?: number;
}

export interface FinanceAmountTriplet {
  quantity?: number;
  pushed_quantity?: number;
  max_push_quantity?: number;
}

export interface DocumentReconciliationGapItem extends FinanceAmountTriplet {
  doc_type?: string;
  doc_id?: number;
  doc_code?: string;
  amount?: number;
  remaining_amount?: number;
  unsettled_amount?: number;
  finance_related_count?: number;
  settlement_type?: string;
  gap_reason?: string | null;
  tree_level?: number;
  parent_doc_id?: number;
  parent_doc_code?: string;
  sort_date?: string;
}

export interface DocumentReconciliationGapResult {
  items?: DocumentReconciliationGapItem[];
  total?: number;
  gap_count?: number;
  open_balance_total?: number;
  partner_type?: string;
  partner_id?: number;
  period?: { start?: string; end?: string };
}

export interface FinancePipelineSummary {
  open_receivable_amount?: number;
  open_payable_amount?: number;
  unsettled_receipt_amount?: number;
  unsettled_payment_amount?: number;
  open_receivable_count?: number;
  open_payable_count?: number;
  unsettled_receipt_count?: number;
  unsettled_payment_count?: number;
  open_finance_document_count?: number;
  tip?: string;
}

const API = '/apps/kuaicaiwu/document-reconciliation';

export interface ChainDocumentCandidate {
  id: number;
  code?: string;
  partner_name?: string;
  label?: string;
}

export const documentReconciliationService = {
  listOpenGaps: (params: DocumentReconciliationGapParams) =>
    apiRequest<DocumentReconciliationGapResult>(`${API}/gaps/open`, { method: 'GET', params }),

  listChainCandidates: (params: { document_type: string; keyword?: string; limit?: number }) =>
    apiRequest<{ items?: ChainDocumentCandidate[]; total?: number }>(`${API}/chain-candidates`, {
      method: 'GET',
      params,
    }),

  getStandardChain: (flowType: 'sales' | 'purchase', documentType: string, documentId: number) =>
    apiRequest<Record<string, unknown>>(`${API}/chain/${flowType}/${documentType}/${documentId}`, { method: 'GET' }),
};

File: riveredge-backend/src/apps/kuaicaiwu/api/document_reconciliation.py
"""
业财单据对账 API
"""

from datetime import date
from typing import Any, Optional

from fastapi import APIRouter, Depends, HTTPException, Query

from apps.kuaicaiwu.api._kuaicaiwu_route_access import require_kuaicaiwu_module_access
from apps.kuaicaiwu.services.document_reconciliation_service import DocumentReconciliationService
from core.api.deps.deps import get_current_user

router = APIRouter(
    prefix="/document-reconciliation",
    tags=["App - Kuaicaiwu - Document Reconciliation"],
    dependencies=[Depends(require_kuaicaiwu_module_access("document-reconciliation"))],
)
service = DocumentReconciliationService()


@router.get("/gaps/open", summary="Open finance gaps by partner and period")
async def list_open_gaps(
    partner_type: str = Query(..., description="Customer 或 Supplier"),
    partner_id: int = Query(...),
    start_date: date = Query(...),
    end_date: date = Query(...),
    only_gaps: bool = Query(True, description="仅返回未关联业财链或仍有未结清金额的单据"),
    keyword: Optional[str] = Query(None),
    doc_type: Optional[str] = Query(None),
    doc_code: Optional[str] = Query(None),
    sort_field: Optional[str] = Query(None),
    sort_order: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=200),
    current_user: Any = Depends(get_current_user),
):
    return await service.list_open_finance_gaps(
        current_user.tenant_id,
        partner_type=partner_type,
        partner_id=partner_id,
        start_date=start_date,
        end_date=end_date,
        only_gaps=only_gaps,
        keyword=keyword,
        doc_type=doc_type,
        doc_code=doc_code,
        sort_field=sort_field,
        sort_order=sort_order,
        skip=skip,
        limit=limit,
    )


@router.get("/pipeline-summary", summary="Tenant finance pipeline summary (read-only)")
async def get_pipeline_summary(current_user: Any = Depends(get_current_user)):
    return await service.get_pipeline_summary(current_user.tenant_id)


@router.get("/chain-candidates", summary="Search chain start documents by code")
async def list_chain_candidates(
    document_type: str = Query(..., description="起始单据类型"),
    keyword: Optional[str] = Query(None, description="单据编号或往来单位名称"),
    limit: int = Query(20, ge=1, le=50),
    current_user: Any = Depends(get_current_user),
):
    from infra.exceptions.exceptions import ValidationError

    try:
        return await service.list_chain_document_candidates(
            current_user.tenant_id,
            document_type=document_type,
            keyword=keyword,
            limit=limit,
        )
    except ValidationError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@router.get("/chain/{flow_type}/{document_type}/{document_id}", summary="Standard sales/purchase finance chain")
async def get_standard_chain(
    flow_type: str,
    document_type: str,
    document_id: int,
    current_user: Any = Depends(get_current_user),
):
    if flow_type.lower() not in ("sales", "purchase"):
        raise HTTPException(status_code=400, detail="flow_type 必须为 sales 或 purchase")
    return await service.build_standard_chain(
        current_user.tenant_id,
        flow_type=flow_type,
        document_type=document_type,
        document_id=document_id,
    )


@router.get("/prepayment-balances", summary="Prepayment balance summary")
async def get_prepayment_balances(
    partner_type: Optional[str] = Query(None, description="customer 或 supplier；传入时分页返回 items"),
    keyword: Optional[str] = Query(None),
    partner_name: Optional[str] = Query(None),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=200),
    sort_field: Optional[str] = Query(None),
    sort_order: Optional[str] = Query(None),
    current_user: Any = Depends(get_current_user),
):
    return await service.get_prepayment_balances(
        current_user.tenant_id,
        partner_type=partner_type,
        keyword=keyword,
        partner_name=partner_name,
        skip=skip,
        limit=limit,
        sort_field=sort_field,
        sort_order=sort_order,
        operator_id=getattr(current_user, "id", None),
    )


@router.get("/{document_type}/{document_id}", summary="Reconcile single document chain")
async def reconcile_document(
    document_type: str,
    document_id: int,
    current_user: Any = Depends(get_current_user),
):
    try:
        return await service.reconcile_document(
            current_user.tenant_id, document_type, document_id
        )
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc))
